psydac/fem/tests/test_dirichlet_projectors.py

```diff
@@ -0,0 +1,562 @@
+    derham_h = discretize(derham, domain_h, degree=degree)
+
+    projectors = derham_h.projectors(nquads=[(d + 1) for d in degree])
+
+    d_projectors = derham_h.dirichlet_projectors(kind='linop')
+
+    nn = NormalVector('nn')
+
+    for i in range(2):
+        print(f'      - Test DBP{i}')
+
+        u, v = elements_of(derham.spaces[i], names='u, v')
+
+        if i == 0:
+            boundary_expr = u*v
+            expr = u*v
+        if (i == 1):
+            boundary_expr = cross(nn, u) * cross(nn, v)
+            expr = inner(u,v)
+
+        a   = BilinearForm((u, v), integral(domain,            expr))            
+        ab  = BilinearForm((u, v), integral(boundary, boundary_expr))
+
+        ah  = discretize(a,  domain_h, (derham_h.spaces[i], derham_h.spaces[i]), backend=backend)
+        abh = discretize(ab, domain_h, (derham_h.spaces[i], derham_h.spaces[i]), backend=backend, sum_factorization=False)
+
+        I   = IdentityOperator(derham_h.spaces[i].coeff_space)
+        DP = d_projectors[i]
+
+        M   = ah.assemble()
+        M_0 = DP @ M @ DP + (I - DP)
+        Mb  = abh.assemble()
+
+        f   = funs[i]
+        fc  = projectors[i](f).coeffs
+
+        # 1.
+        # The coefficients of functions satisfying homogeneous Dirichlet 
+        # boundary conditions should not change under application of the corresponding projector
+        fc2  = DP @ fc
+        diff = fc - fc2
+        err  = diff.inner(diff)
+        print(f' | f - P @ f |^2          = {err}')
+        assert err < tol**2
+
+        # 2.1
+        # After applying a projector to a random vector, we want to verify that the 
+        # corresponding boundary integral vanishes
+        rdm_coeffs = derham_h.spaces[i].coeff_space.zeros()
+        print(' Random boundary integrals:')
+        for _ in range(3):
+            for patch in rdm_coeffs.blocks:
+
+                if isinstance(patch.space, BlockVectorSpace):
+                    for block in patch.blocks:
+                        rng.random(size=block._data.shape, dtype="float64", out=block._data)
+                else:
+                    rng.random(size=patch._data.shape, dtype="float64", out=patch._data)
+
+            rdm_coeffs2 = DP @ rdm_coeffs
+            boundary_int_rdm      = Mb.dot_inner(rdm_coeffs, rdm_coeffs) / rdm_coeffs.space.dimension**2
+            boundary_int_proj_rdm = Mb.dot_inner(rdm_coeffs2, rdm_coeffs2) / rdm_coeffs.space.dimension**2
+            print(f'  rdm: {boundary_int_rdm}    proj. rdm: {boundary_int_proj_rdm}')
+            assert boundary_int_proj_rdm < tol**2
+
+        # 2.2
+        # Test toarray(): (DP @ rdm_coeffs).toarray() should be equal to DP.toarray().dot(rdm_coeffs.toarray())
+        DP_arr = DP.toarray()
+        rdm_coeffs_arr = rdm_coeffs.toarray()
+        diff_arr = DP_arr.dot(rdm_coeffs_arr) - rdm_coeffs2.toarray()
+        err = np.linalg.norm(diff_arr)
+        assert err < tol**2
```

Comment:
Same comment as above
```suggestion
        err = np.linalg.norm(diff_arr)
        assert err < tol
```

Reply:
ff3a5b23e3a623b36a8d635603c617157b49722b